ci/sync-gateway-api.sh

```diff
@@ -0,0 +1,192 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#
+# Gateway API ${CAPITALIZED_CHANNEL} channel install
+#
+---
+EOF
+)
+
+# Function to compare files ignoring headers (for TCPRoute only)
+compare_files_no_header() {
+    local file1="$1"
+    local file2="$2"
+
+    # Strip the header from both files and compare the rest of the content
+    tail -n +$(($(echo "$HEADER" | wc -l) + 1)) "$file1" > "$file1.stripped"
+    tail -n +$(($(echo "$HEADER" | wc -l) + 1)) "$file2" > "$file2.stripped"
+    cmp -s "$file1.stripped" "$file2.stripped"
+    local result=$?
+    rm -f "$file1.stripped" "$file2.stripped"
+    return $result
+}
+
+download_gateway_crd() {
+    local url="$1"
+    local dest="$2"
+
+    echo "Downloading Gateway CRD from $url to $dest..."
+    curl -sLo "$dest.tmp" "$url"
+
+    if [ -f "$dest" ] && cmp -s "$dest" "$dest.tmp"; then
+        echo "No changes detected in $dest."
+        rm "$dest.tmp"
+    else
+        mv "$dest.tmp" "$dest"
+        echo "Updated $dest."
+    fi
+}
+
+download_tcproute_crd() {
+    local url="$1"
+    local dest="$2"
+
+    echo "Downloading TCPRoute CRD from $url to $dest..."
+    curl -sLo "$dest.tmp" "$url"
+
+    # Always create the temporary file with the header
+    echo "$HEADER" > "$dest.tmp.full"
+    cat "$dest.tmp" >> "$dest.tmp.full"
+
+    if [ -f "$dest" ]; then
+        # Compare files ignoring the header
+        if compare_files_no_header "$dest" "$dest.tmp.full"; then
+            echo "No changes detected in $dest."
+            rm "$dest.tmp" "$dest.tmp.full"
+            return
+        fi
+    fi
+
+    # Update the file with the new content and header
+    mv "$dest.tmp.full" "$dest"
+    rm "$dest.tmp"
+    echo "Updated $dest."
+}
+
+# Update sigs.k8s.io/gateway-api in go.mod
+update_go_mod() {
+    local module="sigs.k8s.io/gateway-api"
+    if grep -q "$module" go.mod; then
+        current_version=$(grep "$module" go.mod | awk '{print $2}')
+        if [ "$current_version" != "$CONFORMANCE_VERSION" ]; then
+            echo "Updating $module from $current_version to $CONFORMANCE_VERSION..."
+            go get "$module@$CONFORMANCE_VERSION"
+            go mod tidy
+            echo "Updated $module to $CONFORMANCE_VERSION."
+        else
+            echo "$module is already at version $CONFORMANCE_VERSION."
+        fi
+    else
+        echo "$module not found in go.mod."
+    fi
+}
+
+# Update k8sgateway_api_version in nightly-tests/max_versions.env
+update_max_versions_env() {
```

Comment:
Just thinking aloud -- do we need to handle the min_versions.env file as well? I don't think we've outlined any process around GW API version matrix that gloo supports (e.g. N-1), so automating that file could be premature.